exporters/trace/ocagent/src/main/java/io/opencensus/exporter/trace/ocagent/OcAgentTraceServiceExportRpcHandler.java

```
@@ -0,0 +1,138 @@
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+import io.opencensus.proto.agent.trace.v1.ExportTraceServiceRequest;
+import io.opencensus.proto.agent.trace.v1.ExportTraceServiceResponse;
+import io.opencensus.proto.agent.trace.v1.TraceServiceGrpc.TraceServiceStub;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Handler of export service RPC. */
+@ThreadSafe
+final class OcAgentTraceServiceExportRpcHandler {
+
+  private static final Logger logger =
+      Logger.getLogger(OcAgentTraceServiceExportRpcHandler.class.getName());
+
+  // A reference to the exportRequestObserver returned from stub.
+  @GuardedBy("this")
+  @Nullable
+  private StreamObserver<ExportTraceServiceRequest> exportRequestObserver;
+
+  // The RPC status when this stream finishes/disconnects. Null if the stream is still connected.
+  @GuardedBy("this")
+  @Nullable
+  private Status terminateStatus;
+
+  private OcAgentTraceServiceExportRpcHandler() {}
+
+  private synchronized void setExportRequestObserver(
+      StreamObserver<ExportTraceServiceRequest> exportRequestObserver) {
+    this.exportRequestObserver = exportRequestObserver;
+  }
+
+  // Creates an OcAgentTraceServiceExportRpcHandler. Tries to initiate the export stream with the
+  // given TraceServiceStub.
+  static OcAgentTraceServiceExportRpcHandler create(TraceServiceStub stub) {
+    OcAgentTraceServiceExportRpcHandler exportRpcHandler =
+        new OcAgentTraceServiceExportRpcHandler();
+    ExportResponseObserver exportResponseObserver = new ExportResponseObserver(exportRpcHandler);
+    try {
+      StreamObserver<ExportTraceServiceRequest> exportRequestObserver =
+          stub.export(exportResponseObserver);
+      exportRpcHandler.setExportRequestObserver(exportRequestObserver);
+    } catch (Exception e) {
```

It is not good to catch all Exceptions. What exception do we catch here? Also instead of returning an object should we just throw the exception? That will imply that  "exportRequestObserver" is never null for example.

---

> It is not good to catch all Exceptions. What exception do we catch here?

Make sense, here we should only catch the gRPC `StatusRuntimeException`.

> Also instead of returning an object should we just throw the exception?

I don't think we need to throw the `StatusRuntimeException`, since we've stored the terminate status if we saw a `StatusRuntimeException` and the worker thread can make use of the status. Agree other runtime exceptions can be just thrown to the worker thread. 

> That will imply that "exportRequestObserver" is never null for example.

IMO that's less an issue. `exportRequestObserver` is only created when `exportHandler` is created, and when there's an exception, `exportHandler` will be GC'ed, so is the `exportRequestObserver`.